pkg/koordlet/metricsadvisor/collectors/performance/performance_collector_linux.go

```diff
@@ -79,8 +84,25 @@ func (p *performanceCollector) Run(stopCh <-chan struct{}) {
 	if features.DefaultKoordletFeatureGate.Enabled(features.PSICollector) {
 		p.collectPSI(stopCh)
 	}
-	if features.DefaultKoordletFeatureGate.Enabled(features.CPICollector) {
-		go wait.Until(p.collectContainerCPI, p.cpiCollectInterval, stopCh)
+	if p.EnabledPerf() {
+		if features.DefaultKoordletFeatureGate.Enabled(features.Libpfm4) {
```

Comment:
I'm not sure about this logic here: If both CPICollector and Libpfm4 feature-gate are enabeld, use Libpfm4 to collect CPI, if only CPICollector is enabled, use perf-util way. So should it be 

`
if features.DefaultKoordletFeatureGate.Enabled(features.Libpfm4) {
...
} else {
...
}
`
Or I misunderstood you? @bowen-intel

Reply:
> I'm not sure about this logic here: If both CPICollector and Libpfm4 feature-gate are enabeld, use Libpfm4 to collect CPI, if only CPICollector is enabled, use perf-util way. So should it be
> 
> `if features.DefaultKoordletFeatureGate.Enabled(features.Libpfm4) { ... } else { ... }` Or I misunderstood you? @bowen-intel

I use interface and flag to change the `perfCollector`/`perfGroupCollector` in the `getAndStartCollectorOnSingleContainer`, so if {...} else {...} logic is not used in `Run`.

Reply:
> > I'm not sure about this logic here: If both CPICollector and Libpfm4 feature-gate are enabeld, use Libpfm4 to collect CPI, if only CPICollector is enabled, use perf-util way. So should it be
> > `if features.DefaultKoordletFeatureGate.Enabled(features.Libpfm4) { ... } else { ... }` Or I misunderstood you? @bowen-intel
> 
> I use interface and flag to change the `perfCollector`/`perfGroupCollector` in the `getAndStartCollectorOnSingleContainer`, so if {...} else {...} logic is not used in `Run`.

Yes, I think it's ok to decide which collector to use there. And please see my second comment to determine whether the logic controlling libpfm4 initialization needs to be decoupled from the logic collecting CPI : )
cc @zwzhang0107

Reply:
> Here's some of my opinions:
> 
> 1. `p.EnablePerf()` is related to feature-gate `CPICollector`, but in future we may use Perf to collect other metrics, so this function can be deleted.
> 2. Then we use feature-gate `Libpfm4` to controll if we need to init libpfm4 or init buffer pool, etc.
> 3. To init buffer pool of libpfm4, we need to generate a `eventsMap`, and now you use `if features.DefaultKoordletFeatureGate.Enabled(features.CPICollector)`(line 91). I suggest provide a function `getLibpfm4EventMap`. Then if we add new metrics, only this function's implementation need to be modified. Or, a global featuregate map can be added in file performance_collector_linux.go so `getLibpfm4EventMap` can traverse it.
> 
> The resulting code structure looks like below:
> 
> ```
> func (p *performanceCollector) Run(stopCh <-chan struct{}) {
>         ...
> 
>         if features.DefaultKoordletFeatureGate.Enabled(features.Libpfm4) {
>                 ...
>                 eventsMap := p.getLibpfm4EventMap()
>                 perf.InitBufferPool(eventsMap)
>                 ...
>         }
>         
>         if features.DefaultKoordletFeatureGate.Enabled(features.CPICollector) {
>                  ... // p.collectContainerCPI will check Libpfm4 feature-gate again to really decide which way to use to collect CPI(perf-util or libpfm4)
>         }
> }
> 
> func (p *performanceCollector) getLibpfm4EventMap() map[int]struct{} {
> 	...
> }
> ```
> 
> @bowen-intel

1. I think `p.EnablePerf()` can judge whether perf related feature flag used. if I remove it, I also need to check all perf related flags to init `libpfm4` and events map.
2. `Libpfm4` flag is a sub feature gate for perf related flag. If user try flag `Libpfm4` but forget to use `CPI... `, just ignore it maybe a good choice?
3. Good idea! If I understand your meaning, you want to extract these logic like this?
```go
func (p *performanceCollector) getLibpfm4EventMap() map[int]struct{} {
	eventsMap := make(map[int]struct{})
	if features.DefaultKoordletFeatureGate.Enabled(features.CPICollector) {
		eventsMap[len(perf.EventsMap[string(features.CPICollector)])] = struct{}{}
	}
	return eventsMap
}
```

Reply:
> 3. Good idea! If I understand your meaning, you want to extract these logic like this?

Yes it is. Then if more perf related feature gates are enabled, we can only modify here to add more entries in eventsMap. 

> 1. I think p.EnablePerf() can judge whether perf related feature flag used. if I remove it, I also need to check all perf related flags to init libpfm4 and events map. 
> 2. Libpfm4 flag is a sub feature gate for perf related flag. If user try flag Libpfm4 but forget to use CPI... , just ignore it maybe a good choice?

So you are considering that if user only enable `Libpfm4 `flag but not enable `CPICollector`, to initialize libpfm4 is kind of a waste. To use `p.EnablePerf()` is a double-check to ensure this won't happen? 

If so, just to confirm, if another perf related collector is added in the future, we need to and a OR condition in `p.EnablePerf() `return value. If I'm right, I think it's an acceptable implementation. And this PR will be LGTM after `getLibpfm4EventMap()` function is added. 
@bowen-intel 
cc @zwzhang0107